File: frontend/src/types/schemas.ts
// Mirror of backend/schemas.py. Field names and ranges must match exactly so
// the frontend can consume real pipeline output without translation.

export interface ParsedHypothesis {
  claim: string;
  mechanism: string;
  context: string;
  population: string;
  method: string;
}

export interface Paper {
  paper_id: string;
  title: string;
  authors: string[];
  year: number | null;
  abstract: string;
  url: string;
  citation_count: number;
  relevance_score: number;
  cluster: string;
  doi?: string;
  openalex_id?: string;
  semantic_scholar_id?: string;
  source_provenance?: string[];
}

export interface SourceError {
  source: string;
  query: string;
  error: string;
}

export interface RetrievalReport {
  status: "ok" | "partial" | "no_results" | "failed";
  attempted_queries: string[];
  source_errors: SourceError[];
  cutoff_year?: number | null;
  paper_count: number;
  openalex_total_count: number;
}

export interface Conflict {
  paper_id: string;
  title: string;
  disagreement_dimension: string;
  explanation: string;
  severity: number; // 0..1
}

export interface OverlapReport {
  crowding_score: number; // 0..100
  overlapping_papers: string[];
  whitespace_summary: string;
  risk_notes: string[];
}

export interface ResearchGroup {
  group_id: string;
  name: string;
  institution: string;
  principal_investigators: string[];
  recent_paper_ids: string[];
  methods: string[];
  grounding_evidence: string;
}

export interface EmulatorOutput {
  group_id: string;
  group_name: string;
  interest_score: number; // 0..100
  engagement_type: string;
  proposed_direction: string;
  method_they_use: string;
  time_to_publish_months: number;
  competitive_risk: number; // 0..100
  grounding_paper_ids: string[];
}

export interface Scenario {
  scenario_id: string;
  name: string;
  probability: number; // 0..1
  description: string;
  leading_groups: string[];
  implications: string[];
}

export interface ImpactDimension {
  score: number; // 0..100
  confidence_low: number;
  confidence_high: number;
  rationale: string;
}

export interface ImpactForecast {
  volume: ImpactDimension;
  velocity: ImpactDimension;
  reach: ImpactDimension;
  depth: ImpactDimension;
  disruption: ImpactDimension;
  translation: ImpactDimension;
  overall_summary: string;
}

export type ImpactDimensionKey =
  | "volume"
  | "velocity"
  | "reach"
  | "depth"
  | "disruption"
  | "translation";

export const IMPACT_DIMENSIONS: ImpactDimensionKey[] = [
  "volume",
  "velocity",
  "reach",
  "depth",
  "disruption",
  "translation",
];

export interface MetricScore {
  name: string;
  metric_name?: string;
  score: number;
  confidence_low: number;
  confidence_high: number;
  rationale: string;
  evidence_ids: string[];
  evidence?: string[];
  method?: string;
  weakness?: string;
}

export interface Scorecard {
  composite_score: number;
  verdict: string;
  metric_scores: MetricScore[];
  strengths: string[];
  weaknesses: string[];
  evidence_summary: string;
  metric_weights?: Record<string, number>;
  weighted_contributions?: Record<string, number>;
}

export interface Variant {
  variant_id: string;
  hypothesis_text: string;
  operator: string;
  rationale: string;
  impact_scores: Record<string, number>;
  is_pareto_selected: boolean;
  dominance_explanation: string;
  composite_score?: number;
  rank?: number | null;
  scorecard?: Scorecard | null;
  impact_forecast?: ImpactForecast | null;
}

export interface GroundednessCheck {
  group_id: string;
  group_name: string;
  is_grounded: boolean;
  flagged_inconsistencies: string[];
  evidence: string;
}

export interface StrategyMemo {
  recommendation: string;
  executive_summary: string;
  recommended_hypothesis_text?: string;
  recommended_variant_id?: string | null;
  original_verdict?: string | null;
  recommended_verdict?: string | null;
  scorecard_summary?: string[];
  key_findings: string[];
  selected_variants: string[];
  trade_offs?: string[];
  evidence_trail?: string[];
  risks: string[];
  next_steps: string[];
  denario_next_actions?: string[];
}

// Mirrors backend.pipeline.PipelineState. Frontend-only display extras
// (denario_outline, next_actions, headline_groups) are deliberately separate
// so they don't clash with the backend's authoritative shape.
export interface PipelineState {
  raw_hypothesis: string;
  parsed?: ParsedHypothesis;
  papers?: Paper[];
  conflicts?: Conflict[];
  overlaps?: OverlapReport;
  forecast?: ImpactForecast;
  metric_scores?: MetricScore[];
  scorecard?: Scorecard;
  variants?: Variant[];
  rescored_variants?: Variant[];
  pareto_variants?: Variant[];
  ranked_variants?: Variant[];
  final_memo?: StrategyMemo;
  information_cutoff_year?: number | null;
  backtest_metadata?: Record<string, unknown> | null;
  openalex_total_count?: number | null;
  retrieval_report?: RetrievalReport | null;
}

// Frontend-only additions used for demo presentation; not produced by the backend.
export interface DenarioOutline {
  title: string;
  abstract: string;
  methods: string;
  experiments: string;
  discussion: string;
}

export interface DemoExtras {
  headline_groups: string[]; // group_ids the narrator zooms into
  denario_outline: DenarioOutline;
  next_actions: string[];
}

export interface DemoState extends PipelineState {
  papers: Paper[];
  conflicts: Conflict[];
  groups: ResearchGroup[];
  emulator_outputs: EmulatorOutput[];
  scenarios: Scenario[];
  variants: Variant[];
  groundedness_checks: GroundednessCheck[];
  extras: DemoExtras;
}

File: backend/schemas.py
from __future__ import annotations

from typing import Dict, List, Literal, Optional

from pydantic import AliasChoices, BaseModel, ConfigDict, Field, computed_field


class ParsedHypothesis(BaseModel):
    claim: str
    mechanism: str
    context: str
    population: str
    method: str


class Paper(BaseModel):
    paper_id: str
    title: str
    authors: List[str] = Field(default_factory=list)
    year: Optional[int] = None
    abstract: str = ""
    url: str = ""
    citation_count: int = 0
    relevance_score: float = 0.0
    cluster: str = "unclustered"
    doi: str = ""
    openalex_id: str = ""
    semantic_scholar_id: str = ""
    source_provenance: List[str] = Field(default_factory=list)


class SourceError(BaseModel):
    source: str
    query: str = ""
    error: str


class RetrievalReport(BaseModel):
    status: Literal["ok", "partial", "no_results", "failed"]
    attempted_queries: List[str] = Field(default_factory=list)
    source_errors: List[SourceError] = Field(default_factory=list)
    cutoff_year: Optional[int] = None
    paper_count: int = Field(default=0, ge=0)
    openalex_total_count: int = Field(default=0, ge=0)


class Conflict(BaseModel):
    paper_id: str
    title: str
    disagreement_dimension: str
    explanation: str
    severity: float = Field(ge=0.0, le=1.0)


class OverlapReport(BaseModel):
    crowding_score: int = Field(ge=0, le=100)
    overlapping_papers: List[str] = Field(default_factory=list)
    whitespace_summary: str
    risk_notes: List[str] = Field(default_factory=list)


class ResearchGroup(BaseModel):
    group_id: str
    name: str
    institution: str
    principal_investigators: List[str] = Field(default_factory=list)
    recent_paper_ids: List[str] = Field(default_factory=list)
    methods: List[str] = Field(default_factory=list)
    grounding_evidence: str


class EmulatorOutput(BaseModel):
    group_id: str
    group_name: str
    interest_score: int = Field(ge=0, le=100)
    engagement_type: str
    proposed_direction: str
    method_they_use: str
    time_to_publish_months: int = Field(ge=0)
    competitive_risk: int = Field(ge=0, le=100)
    grounding_paper_ids: List[str] = Field(default_factory=list)


class Scenario(BaseModel):
    scenario_id: str
    name: str
    probability: float = Field(ge=0.0, le=1.0)
    description: str
    leading_groups: List[str] = Field(default_factory=list)
    implications: List[str] = Field(default_factory=list)


class ImpactDimension(BaseModel):
    score: int = Field(ge=0, le=100)
    confidence_low: int = Field(ge=0, le=100)
    confidence_high: int = Field(ge=0, le=100)
    rationale: str
    predicted_value: Optional[float] = None
    predicted_low: Optional[float] = None
    predicted_high: Optional[float] = None
    units: str = ""


class ImpactForecast(BaseModel):
    volume: ImpactDimension
    velocity: ImpactDimension
    reach: ImpactDimension
    depth: ImpactDimension
    disruption: ImpactDimension
    translation: ImpactDimension
    overall_summary: str


class MetricScore(BaseModel):
    """Quantitative score for one Idea Hater metric.

    `name` and `evidence_ids` are the canonical fields. Legacy aliases
    `metric_name` and `evidence` are accepted to keep the current pipeline and
    dashboard stubs working while the backend moves away from group emulation.
    """

    model_config = ConfigDict(populate_by_name=True)

    name: str = Field(validation_alias=AliasChoices("name", "metric_name"))
    score: int = Field(ge=0, le=100)
    confidence_low: int = Field(ge=0, le=100)
    confidence_high: int = Field(ge=0, le=100)
    rationale: str
    evidence_ids: List[str] = Field(
        default_factory=list,
        validation_alias=AliasChoices("evidence_ids", "evidence"),
    )
    method: str = ""
    weakness: str = ""

    @computed_field
    @property
    def metric_name(self) -> str:
        return self.name

    @computed_field
    @property
    def evidence(self) -> List[str]:
        return self.evidence_ids


class IdeaEvaluation(BaseModel):
    """Full quantitative scorecard for an original hypothesis or variant."""

    hypothesis_text: str
    parsed: ParsedHypothesis
    metrics: List[MetricScore] = Field(default_factory=list)
    composite_score: int = Field(ge=0, le=100)
    verdict: Literal["reject", "revise", "promising", "strong"]
    key_weaknesses: List[str] = Field(default_factory=list)
    key_strengths: List[str] = Field(default_factory=list)
    metric_weights: Dict[str, float] = Field(default_factory=dict)
    evidence_summary: str = ""


class Scorecard(BaseModel):
    """Display-oriented aggregate scorecard used by the current pipeline."""

    composite_score: int = Field(ge=0, le=100)
    verdict: str
    metric_scores: List[MetricScore] = Field(default_factory=list)
    strengths: List[str] = Field(default_factory=list)
    weaknesses: List[str] = Field(default_factory=list)
    evidence_summary: str
    metric_weights: Dict[str, float] = Field(default_factory=dict)
    weighted_contributions: Dict[str, float] = Field(default_factory=dict)


class Variant(BaseModel):
    variant_id: str
    hypothesis_text: str
    operator: str
    rationale: str
    impact_scores: Dict[str, int] = Field(default_factory=dict)

    impact_forecast: Optional[ImpactForecast] = None

    composite_score: int = Field(default=0, ge=0, le=100)
    scorecard: Optional[Scorecard] = None
    rank: Optional[int] = None

    is_pareto_selected: bool = False
    dominance_explanation: str = ""


class VariantEvaluation(BaseModel):
    """Evaluation and Pareto status for a generated hypothesis variant."""

    variant: Variant
    evaluation: IdeaEvaluation
    improvement_over_original: int
    is_pareto_selected: bool = False
    dominance_explanation: str = ""


class GroundednessCheck(BaseModel):
    group_id: str
    group_name: str
    is_grounded: bool
    flagged_inconsistencies: List[str] = Field(default_factory=list)
    evidence: str


class StrategyMemo(BaseModel):
    """Final Denario-facing recommendation built from scores and evidence."""

    recommendation: str
    executive_summary: str
    recommended_hypothesis_text: str = ""
    recommended_variant_id: Optional[str] = None
    original_verdict: Optional[str] = None
    recommended_verdict: Optional[str] = None
    scorecard_summary: List[str] = Field(default_factory=list)
    key_findings: List[str] = Field(default_factory=list)
    selected_variants: List[str] = Field(default_factory=list)
    trade_offs: List[str] = Field(default_factory=list)
    evidence_trail: List[str] = Field(default_factory=list)
    risks: List[str] = Field(default_factory=list)
    next_steps: List[str] = Field(default_factory=list)
    denario_next_actions: List[str] = Field(default_factory=list)

    def to_markdown(self) -> str:
        sections = [
            "# Strategy Memo",
            f"## Recommendation\n{self.recommendation}",
            f"## Recommended Hypothesis\n{self.recommended_hypothesis_text}" if self.recommended_hypothesis_text else "",
            f"## Executive Summary\n{self.executive_summary}",
            _bullets("Scorecard Summary", self.scorecard_summary),
            _bullets("Key Findings", self.key_findings),
            _bullets("Selected Variants", self.selected_variants),
            _bullets("Trade-offs", self.trade_offs),
            _bullets("Evidence Trail", self.evidence_trail),
            _bullets("Risks", self.risks),
            _bullets("Next Steps", self.next_steps),
            _bullets("Denario Next Actions", self.denario_next_actions),
        ]
        return "\n\n".join(section for section in sections if section)


def _bullets(title: str, values: List[str]) -> str:
    if not values:
        return ""
    body = "\n".join(f"- {value}" for value in values)
    return f"## {title}\n{body}"
